### ISSUE TITLE
Slow training

### ISSUE THREAD

**Issue by ashafiez** [NextEasywood/Model-Based-Deep-Learning-for-Low-Cost-IMU-Dead-Reckoning-of-Wheeled-Mobile-Robot#9]
Hi, Thank you for the valuable work, the results look very promising. 
I would like to try to training on the kitti sequences (unsync with IMU 100Hz), but the training process seems to be very slow. Each epoch takes about 20-30 seconds to be completed. Is that normal? If no, do you have an idea why? And if yes, do you have an idea how we can improve it to be faster? 
Thanks in advance for your help. 

Below is an snapshot of the training (pre train and then main train loop): 
Train Epoch: 3999       Loss: 1.1105
Train Epoch: 4000       Loss: 1.1211
Amount of time spent for epochs 3950-4000: 4.4s

validation loss decreases! :) (curr/prev loss 0.9262/0.9242)
('2011_09_26_drive_0022_extract', '2011_09_29_drive_0071_extract', '2011_09_30_drive_0018_extract', '2011_09_30_drive_0020_extract', '2011_09_30_drive_0027_extract', '2011_09_30_drive_0028_extract', '2011_10_03_drive_0027_extract', '2011_10_03_drive_0034_extract', '2011_10_03_drive_0047_extract') train_time_Loss =  16.583s
Train Epoch:  1         Loss: 6.0464
('2011_09_26_drive_0022_extract', '2011_09_29_drive_0071_extract', '2011_09_30_drive_0018_extract', '2011_09_30_drive_0020_extract', '2011_09_30_drive_0027_extract', '2011_09_30_drive_0028_extract', '2011_10_03_drive_0027_extract', '2011_10_03_drive_0034_extract', '2011_10_03_drive_0047_extract') train_time_Loss =  38.002s
Train Epoch:  2         Loss: 72.1001

**Comment by NextEasywood** [comment]
Hi, thanks for the kind words — glad the results look promising.

About the speed: **the per-epoch time is strongly hardware-dependent**, so some fluctuation is normal (CPU model, single-core performance, memory bandwidth, disk speed, and whether you’re actually using a GPU can all change the timing). That said, **this codebase is indeed relatively slow**.

The main reason is that during backpropagation it needs to compute **both**:

1. the neural network gradients, and
2. the **InEKF-related gradient/Jacobian information** used in the optimization.

So the bottleneck is often not “just GPU compute”; a CPU with **strong single-core performance** can help, because parts of the pipeline are effectively CPU-bound / not perfectly parallelized.

Honestly, I don’t have a better practical trick to significantly speed it up right now. In my previous experiments with this same code, training also typically took **more than a day** to finish.
